### PR TITLE
feature/CP-4039-android-geo-fences-with-time-limits_v1

### PR DESCRIPTION
**Pull request by cpKevinrodriguesnobre**
CP-4039 - not getting state trantition state

### CODE REVIEW

File: cleverpush/src/main/java/com/cleverpush/CleverPush.java [hunk 1]
```diff
@@ -1022,8 +1023,10 @@ public void onConnected(@Nullable Bundle bundle) {
                                 geofencePendingIntent
```
Comment (cpKevinrodriguesnobre) [inline cleverpush/src/main/java/com/cleverpush/CleverPush.java:1023]: L1008 where you call addGeofences
maybe do it like here and see if it helps
so add the pendingIntent to the parameters of addGeofences and add the success and failurelistener and ofc some logs inside those
@ozamanav  @jprajapati1996 

image

Comment (cpKevinrodriguesnobre) [reply cleverpush/src/main/java/com/cleverpush/CleverPush.java:1023]: https://developer.android.com/training/location/geofencing#define-a-broadcast-receiver-for-geofence-transitions

File: cleverpush/src/main/java/com/cleverpush/CleverPush.java [hunk 2]
```diff
@@ -962,6 +962,7 @@ void initGeoFences() {
                                                     geoFence.getDouble("longitude"),
                                                     geoFence.getLong("radius")
                                             )
+                                            .setLoiteringDelay((int) geoFence.getDouble("delay"))
```
[the remark this answers is not in the document]
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/CleverPush.java:965]: okay i will comment this line

File: cleverpush/src/main/java/com/cleverpush/CleverPush.java (hunk 1 above)
Comment (cpKevinrodriguesnobre) [reply cleverpush/src/main/java/com/cleverpush/CleverPush.java:1074]: and does it work now? or what is the situation now?

Comment (cpKevinrodriguesnobre) [reply cleverpush/src/main/java/com/cleverpush/CleverPush.java:1074]: also you changed way more than i was asking for and the pendingintent already existed here https://github.com/cleverpush/cleverpush-android-sdk/pull/85/commits/d53e6baa6e9f3fa8da9bba5af0928f9d15cb9168#diff-dadb76999f5ed7d5f6c363d011ef4b354183a8b5cee4ecd118f2152c882891f1R1068
please check with manav if you didnt yet

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 3]
```diff
@@ -22,51 +23,55 @@
...
+
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
+
+        String channelId = sharedPreferences.getString(CleverPushPreferences.CHANNEL_ID, null);
+        String subscriptionId = sharedPreferences.getString(CleverPushPreferences.SUBSCRIPTION_ID, null);
+        String transitionState = event.getGeofenceTransition() == Geofence.GEOFENCE_TRANSITION_ENTER ? "enter" : "exit";
+        Logger.d("transmission", String.valueOf(event.getGeofenceTransition()));
+
+
+//        Logger.d(TAG, "Geofence Transition Details: " + getGeofenceTransitionDetails(event) + " " + transitionState + " subscription: " + subscriptionId + " channel " + channelId);
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:49]: Remove comments

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 4]
```diff
@@ -22,51 +23,55 @@
...
+
+        if (channelId != null && subscriptionId != null) {
+//			for (Geofence geofence : event.getTriggeringGeofences()) {
+            JSONObject jsonBody = new JSONObject();
+            try {
+//					jsonBody.put("geoFenceId", geofence.getRequestId());
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:55]: Remove comments

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 5]
```diff
@@ -22,51 +23,55 @@
...
+        String channelId = sharedPreferences.getString(CleverPushPreferences.CHANNEL_ID, null);
+        String subscriptionId = sharedPreferences.getString(CleverPushPreferences.SUBSCRIPTION_ID, null);
+        String transitionState = event.getGeofenceTransition() == Geofence.GEOFENCE_TRANSITION_ENTER ? "enter" : "exit";
+        Logger.d("transmission", String.valueOf(event.getGeofenceTransition()));
+
+
+//        Logger.d(TAG, "Geofence Transition Details: " + getGeofenceTransitionDetails(event) + " " + transitionState + " subscription: " + subscriptionId + " channel " + channelId);
+
+        if (channelId != null && subscriptionId != null) {
+//			for (Geofence geofence : event.getTriggeringGeofences()) {
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:52]: Remove comments

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 6]
```diff
@@ -22,51 +23,55 @@
...
+                jsonBody.put("subscriptionId", subscriptionId);
+                jsonBody.put("state", transitionState);
+
+                CleverPushHttpClient.post("/subscription/geo-fence", jsonBody, null);
+            } catch (JSONException e) {
+                Logger.e(LOG_TAG, "Error generating geo-fence json", e);
+            }
+        }
+    }
+//	}
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:67]: Remove comments + new line

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 7]
```diff
@@ -22,51 +23,55 @@
...
-		List<String> triggeringIDs = new ArrayList<>();
-		for (Geofence geofence : event.getTriggeringGeofences()) {
-			triggeringIDs.add(geofence.getRequestId());
-		}
-		return String.format("%s: %s", transitionString, TextUtils.join(", ", triggeringIDs));
-	}
+    protected static final String TAG = "CPGeofenceTransitionsIS";
+
+    public CleverPushGeofenceTransitionsIntentService() {
+        super(TAG);  // use TAG to name the IntentService worker thread
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:29]: Remove comments + new line

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 8]
```diff
@@ -22,51 +23,55 @@
...
+    }
+
+    @Override
+    protected void onHandleIntent(Intent intent) {
+//        Logger.d("onHandleIntent","onHandleIntent");
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:34]: Remove comments + new line

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                        mCompleted = false;
+                        cTimer.cancel();
+                        position += 1;
+                        countDownCalling(geofencingEvent);
+                    }
+                }
+            }.start();
+        }
+    }
+}
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:115]: add new line at EOF

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                                public void onSuccess(String response) {
+                                }
+
+                                @Override
+                                public void onFailure(int statusCode, String response, Throwable throwable) {
+                                }
+                            });
+                        }
+                    } else {
+                        Log.e("CPJSONInvalid", "Invalid type");
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:102]: Use `Logger` class

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    private String transitionState;
+    int position;
+    boolean mCompleted = false;
+    CountDownTimer cTimer;
+
+    public void onReceive(Context context, Intent intent) {
+        GeofencingEvent geofencingEvent = GeofencingEvent.fromIntent(intent);
+        if (geofencingEvent.hasError()) {
+            String errorMessage = GeofenceStatusCodes.getStatusCodeString(geofencingEvent.getErrorCode());
+            Log.e("CPErrorMessage", errorMessage);
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:38]: Use `Logger` class and no "CPErrorMessage needed

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+public class GeofenceBroadcastReceiver extends BroadcastReceiver {
+    private double delay;
+    private final ArrayList<Double> mDelayList = new ArrayList();
+    private int transition;
+    private String channelId;
+    private String subscriptionId;
+    private String transitionState;
+    int position;
+    boolean mCompleted = false;
+    CountDownTimer cTimer;
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:32]: Rename to countDownTimer

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                    Logger.d("Exception", ex.getMessage());
+                }
+            }
+        });
+        countDownCalling(geofencingEvent);
+    }
+    
+    void countDownCalling(GeofencingEvent geofencingEvent) {
+        if (position != mDelayList.size()) {
+            cTimer = new CountDownTimer((long) (mDelayList.get(position) * 1000), 1000) {
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:71]: magic number

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                @Override
+                public void onTick(long l) {
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:73]: onTick not used - needs to be overridden?

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                        List<Geofence> triggeringGeofence = geofencingEvent.getTriggeringGeofences();
+                        if (channelId != null && subscriptionId != null && position < triggeringGeofence.size()) {
+                            JSONObject jsonBody = new JSONObject();
+                            try {
+                                jsonBody.put("geoFenceId", triggeringGeofence.get(position).getRequestId());
+                                jsonBody.put("channelId", channelId);
+                                jsonBody.put("subscriptionId", subscriptionId);
+                                jsonBody.put("state", transitionState);
+                            } catch (JSONException e) {
+                                Logger.e("CPJSONException", "Error generating geo-fence json", e);
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:88]: Remove string CPJSONException

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+                            }
+                            mCompleted = true;
+                            CleverPushHttpClient.post("/subscription/geo-fence", jsonBody, new CleverPushHttpClient.ResponseHandler() {
+                                @Override
+                                public void onSuccess(String response) {
+                                }
+
+                                @Override
+                                public void onFailure(int statusCode, String response, Throwable throwable) {
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:97]: Add Logger.e log here for failure

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+                                jsonBody.put("channelId", channelId);
+                                jsonBody.put("subscriptionId", subscriptionId);
+                                jsonBody.put("state", transitionState);
+                            } catch (JSONException e) {
+                                Logger.e("CPJSONException", "Error generating geo-fence json", e);
+                            }
+                            mCompleted = true;
+                            CleverPushHttpClient.post("/subscription/geo-fence", jsonBody, new CleverPushHttpClient.ResponseHandler() {
+                                @Override
+                                public void onSuccess(String response) {
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:93]: Add Logger.d debug log here

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        if (geofencingEvent.hasError()) {
+            String errorMessage = GeofenceStatusCodes.getStatusCodeString(geofencingEvent.getErrorCode());
+            Log.e("CPErrorMessage", errorMessage);
+            return;
+        }
+
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+        channelId = sharedPreferences.getString(CleverPushPreferences.CHANNEL_ID, null);
+        subscriptionId = sharedPreferences.getString(CleverPushPreferences.SUBSCRIPTION_ID, null);
+        transitionState = geofencingEvent.getGeofenceTransition() == Geofence.GEOFENCE_TRANSITION_ENTER ? "enter" : "exit";
```
Comment (gbhrdt) [inline cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:45]: Magic strings. Constants or ENUM

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 9 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:115]: Added

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java (hunk 8 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:34]: okay removed comment and added new line

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java (hunk 3 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:49]: removed

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 10 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:102]: okay used Logger class

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 11 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:38]: okay removed Log

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 12 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:32]: okay renamed

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 15 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:88]: okay removed only string

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 16 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:97]: `Logger.e("", response);`
Added

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 17 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:93]: ` Logger.d("onSuccess", response);`
Added

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+                        for (int i = 0; i < geoFenceArray.length(); i++) {
+                            JSONObject geoFence = geoFenceArray.getJSONObject(i);
+                            if (geoFence != null) {
+                                delay = geoFence.getDouble("delay");
+                                mDelayList.add(delay);
+                            }
+                        }
+                    }
+                } catch (Exception ex) {
+                    Logger.d("Exception", ex.getMessage());
```
[the remark this answers is not in the document]
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:62]: Okay changed

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java (hunk 18 above)
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:45]: Added 
` String ENTER = "enter";
    String EXIT = "exit";`

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+public class GeofenceBroadcastReceiver extends BroadcastReceiver {
+    private double delay;
+    private final ArrayList<Double> mDelayList = new ArrayList();
+    private int transition;
+    private String channelId;
+    private String subscriptionId;
+    private String transitionState;
+    int position;
+    boolean mCompleted = false;
```
[the remark this answers is not in the document]
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:31]: `geofenceIndex`

File: cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.List;
+
+public class GeofenceBroadcastReceiver extends BroadcastReceiver {
+    private double delay;
+    private final ArrayList<Double> mDelayList = new ArrayList();
+    private int transition;
+    private String channelId;
+    private String subscriptionId;
+    private String transitionState;
+    int position;
```
[the remark this answers is not in the document]
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/GeofenceBroadcastReceiver.java:30]: `geoFenceTimeoutCompleted`

File: cleverpush/src/main/java/com/cleverpush/CleverPush.java [hunk 22]
```diff
@@ -1001,30 +1023,28 @@ public void onConnected(@Nullable Bundle bundle) {
...
-                        );
-                    } catch (SecurityException securityException) {
-                        // Catch exception generated if the app does not use ACCESS_FINE_LOCATION permission.
-                    }
+                    geofencingClient.addGeofences(getGeofencingRequest(), getGeofencePendingIntent())
+                            .addOnSuccessListener(getCurrentActivity(), new OnSuccessListener<Void>() {
+                                @Override
+                                public void onSuccess(Void aVoid) {
+                                    try {
+/*
```
[the remark this answers is not in the document]
Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/CleverPush.java:1031]: not needed so i removed

**Review by gbhrdt — APPROVED** [review]

**Comment by gbhrdt** [comment]
Waiting for review from @cpKevinrodriguesnobre

File: cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java [hunk 23]
```diff
@@ -22,51 +25,51 @@
...
-					jsonBody.put("geoFenceId", geofence.getRequestId());
-					jsonBody.put("channelId", channelId);
-					jsonBody.put("subscriptionId", subscriptionId);
-					jsonBody.put("state", transitionState);
+        if (channelId != null && subscriptionId != null) {
+            for (Geofence geofence : event.getTriggeringGeofences()) {
+                JSONObject jsonBody = new JSONObject();
+                try {
+                    jsonBody.put("geoFenceId", geofence.getRequestId());
+                    jsonBody.put("geoFenceId", "geofence.getRequestId()");
```
Comment (cpKevinrodriguesnobre) [inline cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:54]: copy paste mistake?

Comment (cpKevinrodriguesnobre) [reply cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:54]: + resolve conflicts

Comment (jprajapati1996) [reply cleverpush/src/main/java/com/cleverpush/service/CleverPushGeofenceTransitionsIntentService.java:54]: Yes sorry it's copy-paste mistake 
and now all the conflicts are resolved

**Review by cpKevinrodriguesnobre — COMMENTED** [review]
Approved

**Pull request merged**